### PR TITLE
Interpret syntax imported from JSON

### PR DESCRIPTION
**Pull request by robrix**
This PR defines a mapping of Python syntax to a language-agnostic graph format which we can then interpret in `semantic-analysis` via a simple JSON schema (defined by `tree-sitter-graph`).

### CODE REVIEW

File: semantic-analysis/src/Analysis/Syntax.hs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  contents <- B.readFile path
+  pure $ snd <$> A.eitherDecodeWith A.json' (A.iparse parseGraph) contents
+
+parseGraph :: Syntax rep => A.Value -> A.Parser (IntMap.IntMap rep, Maybe rep)
+parseGraph = A.withArray "nodes" $ \ nodes -> do
+  (untied, First root) <- foldMap (\ (k, v, r) -> ([(k, v)], First r)) <$> traverse (A.withObject "node" parseNode) (V.toList nodes)
+  let tied = fix (\ tied -> ($ tied) <$> IntMap.fromList untied)
+  pure (tied, ($ tied) <$> root)
+
+parseNode :: Syntax rep => A.Object -> A.Parser (IntMap.Key, IntMap.IntMap rep -> rep, Maybe (IntMap.IntMap rep -> rep))
```
Comment (robrix) [inline semantic-analysis/src/Analysis/Syntax.hs:155]: This type, specifically the type produced by the parser, is frustrating. If `Parser` had a `MonadFix` instance, we'd just return the first-order terms we want. Instead, we have to return a higher-order graph and root node, and then manually take their fixpoint after the fact.

An alternative approach, and one I think I might prefer if and when I need to dive into this further, would be to build up a graph as an `IntMap` passed in and returned, i.e. in `State`–style. Then  you'd have a test at the top level here checking for the object's index in the graph and short-circuiting if it's found. It complicates the mechanism here, but so does all this higher-order nonsense, so what can you do?

File: semantic-analysis/src/Analysis/Syntax.hs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+parseFile :: Syntax rep => FilePath -> IO (Either (A.JSONPath, String) (Maybe rep))
+parseFile path = do
+  contents <- B.readFile path
+  pure $ snd <$> A.eitherDecodeWith A.json' (A.iparse parseGraph) contents
+
+parseGraph :: Syntax rep => A.Value -> A.Parser (IntMap.IntMap rep, Maybe rep)
+parseGraph = A.withArray "nodes" $ \ nodes -> do
+  (untied, First root) <- foldMap (\ (k, v, r) -> ([(k, v)], First r)) <$> traverse (A.withObject "node" parseNode) (V.toList nodes)
+  let tied = fix (\ tied -> ($ tied) <$> IntMap.fromList untied)
+  pure (tied, ($ tied) <$> root)
```
Comment (robrix) [inline semantic-analysis/src/Analysis/Syntax.hs:153]: Here's where we "tie the knot," taking the non-recursive higher-order data and applying it recursively to the result of applying it recursively to the result of… you get where I'm going with this.

We also need to do the same for the root node, which is also defined over the completed graph.

File: tsg [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+(block (_)* @children) @this
+{
+  node @this.node
+  attr (@this.node) type = "block"
+  for child in @children {
+    edge @this.node -> child.node
+  }
+}
```
Comment (robrix) [inline tsg:52]: New! Fancy!™

**Comment by robrix** [comment]
Following review, I'll make the status check thing use the new job name and then merge.

File: semantic-analysis/src/Analysis/Syntax.hs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+-- Abstract interpretation
+
+eval0 :: Interpret m i -> m i
+eval0 = fix eval
+
+eval :: (Interpret m i -> m i) -> (Interpret m i -> m i)
+eval eval (Interpret f) = f eval
+
+newtype Interpret m i = Interpret { interpret :: (Interpret m i -> m i) -> m i }
```
[the remark this answers is not in the document]
Comment (patrickt) [reply semantic-analysis/src/Analysis/Syntax.hs:114]: A question for my edification: how does this differ from the open-recursion we used in our implementation of _Abstracting Definitional Interpreters_?

File: tsg [hunk 5]
```diff
@@ -0,0 +1,80 @@
+(module) @this
```
Comment (patrickt) [inline tsg:1]: Can we put some comments (tsg supports comments, right?) that indicate what this file is for, should someone stumble across it?

File: semantic-analysis/src/Analysis/Syntax.hs [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+  string :: Text -> rep
+
+  throw :: rep -> rep
+
+  let_ :: Name -> rep -> (rep -> rep) -> rep
+
+
+-- Pretty-printing
+
+newtype Print = Print { print_ :: ShowS }
```
[the remark this answers is not in the document]
Comment (patrickt) [reply semantic-analysis/src/Analysis/Syntax.hs:63]: The newtype over `ShowS` is cute.

File: semantic-analysis/src/Analysis/Syntax.hs (hunk 2 above)
Comment (patrickt) [reply semantic-analysis/src/Analysis/Syntax.hs:152]: This is lovely; can you move this explanation into a code comment?

File: semantic-analysis/src/Analysis/Syntax.hs (hunk 1 above)
Comment (patrickt) [reply semantic-analysis/src/Analysis/Syntax.hs:154]: This is needed because of HOAS, right?

File: semantic-analysis/src/Analysis/Syntax.hs [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  let parseType attrs = \case
+        "string" -> const . string <$> attrs A..: pack "text"
+        "true"   -> pure (const (bool True))
+        "false"  -> pure (const (bool False))
+        "throw"  -> fmap throw <$> resolve (head edges)
+        "if"     -> liftA3 iff <$> findEdgeNamed "condition" <*> findEdgeNamed "consequence" <*> findEdgeNamed "alternative" <|> pure (const noop)
+        "block"  -> children
+        "module" -> children
+        t        -> A.parseFail ("unrecognized type: " <> t)
+      children = fmap (foldr (\ (i, v) r -> let_ (nameI i) v (const r)) noop . zip [0..]) . sequenceA <$> traverse resolve edges
```
Comment (patrickt) [inline semantic-analysis/src/Analysis/Syntax.hs:167]: This right here is pretty opaque; can you add a type signature and perhaps make it a little more verbose? Maybe pull out the fold function into something explicitly named?

File: semantic-analysis/src/Analysis/Syntax.hs [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        sink <- edge A..: pack "sink"
+        attrs <- edge A..: pack "attrs"
+        f attrs
+        pure (IntMap.! sink))
+      findEdgeNamed :: (A.FromJSON a, Eq a) => a -> A.Parser (IntMap.IntMap rep -> rep)
+      findEdgeNamed name = foldMap (resolveWith (\ attrs -> attrs A..: pack "type" >>= guard . (== name))) edges
+  o A..: pack "attrs" >>= A.withObject "attrs" (\ attrs -> do
+    ty <- attrs A..: pack "type"
+    node <- parseType attrs ty
+    pure (index, node, if ty == "module" then Just node else Nothing))
```
Comment (patrickt) [inline semantic-analysis/src/Analysis/Syntax.hs:180]: code golf: maybe `node <$ guard (ty == "module")`?

**Review by patrickt — CHANGES_REQUESTED** [review]

File: semantic-analysis/src/Analysis/Syntax.hs (hunk 1 above)
Comment (robrix) [reply semantic-analysis/src/Analysis/Syntax.hs:154]: Nope, just for the graph to be cyclic, and complete. There's no guarantee that a node referenced during importing has already been imported itself; likewise, we can have recursive references, which can never be finished importing before we complete due to self-reference. Returning a map from the complete graph to individual node representations delays these lookups until the graph is available, and therefore requires a fixpoint (above) to tie the knot.

File: semantic-analysis/src/Analysis/Syntax.hs [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+  throw e = parens (str "throw" <+> e)
+
+  let_ n v b = parens (str "let" <+> name n <+> char '=' <+> v <+> str "in" <+> b (name n))
+
+str :: String -> Print
+str = Print . showString
+
+text :: Text -> Print
+text = str . unpack
```
Comment (patrickt) [inline semantic-analysis/src/Analysis/Syntax.hs:83]: You could unify `text` and `str` if you used `IsString`, it occurs to me

**Review by patrickt — APPROVED** [review]
LGTM!

File: semantic-analysis/src/Analysis/Syntax.hs (hunk 9 above)
Comment (robrix) [reply semantic-analysis/src/Analysis/Syntax.hs:83]: True, but I like the explicitness here. Also `Print` is basically debug code, not something I expect will see much use or maintenance 😊

**Pull request merged**